core/common/src/Clock.kt

```diff
@@ -40,17 +40,50 @@ public fun Clock.todayIn(timeZone: TimeZone): LocalDate =
  * Returns a [TimeSource] that uses this [Clock] to mark a time instant and to find the amount of time elapsed since that mark.
  */
 @ExperimentalTime
-public fun Clock.asTimeSource(): TimeSource = object : TimeSource {
-    override fun markNow(): TimeMark = InstantTimeMark(now(), this@asTimeSource)
+public fun Clock.asTimeSource(): TimeSource.WithComparableMarks = object : TimeSource.WithComparableMarks {
+    override fun markNow(): ComparableTimeMark = InstantTimeMark(now(), this@asTimeSource)
 }
 
 @ExperimentalTime
-private class InstantTimeMark(private val instant: Instant, private val clock: Clock) : TimeMark {
-    override fun elapsedNow(): Duration = clock.now() - instant
+private class InstantTimeMark(private val instant: Instant, private val clock: Clock) : ComparableTimeMark {
+    override fun elapsedNow(): Duration = saturatingDiff(clock.now(), instant)
 
-    override fun plus(duration: Duration): TimeMark = InstantTimeMark(instant + duration, clock)
+    override fun plus(duration: Duration): ComparableTimeMark = InstantTimeMark(instant.saturatingAdd(duration), clock)
+    override fun minus(duration: Duration): ComparableTimeMark = InstantTimeMark(instant.saturatingAdd(-duration), clock)
 
-    override fun minus(duration: Duration): TimeMark = InstantTimeMark(instant - duration, clock)
+    override fun minus(other: ComparableTimeMark): Duration {
+        if (other !is InstantTimeMark || other.clock != this.clock) {
+            throw IllegalArgumentException("Subtracting or comparing time marks from different time sources is not possible: $this and $other")
+        }
+        return saturatingDiff(this.instant, other.instant)
+    }
+
+    override fun equals(other: Any?): Boolean {
+        return other is InstantTimeMark && this.clock == other.clock && this.instant == other.instant
+    }
+
+    override fun hashCode(): Int = instant.hashCode()
+
+    override fun toString(): String = instant.toString()
```

Comment:
I'm wary of time marks with different clocks looking the same but not being equal. I don't think anyone will want to parse these, so I don't see any downsides to including the `clock` in the output as well.

Reply:
ok, makes sense